executor/src/witgen/mod.rs

```diff
@@ -549,6 +567,35 @@ impl<'a, T: FieldElement> FixedData<'a, T> {
                 v.get(row as usize).cloned()
             })
     }
+
+    fn current_stage_witnesses(&self) -> impl Iterator<Item = PolyID> + '_ {
+        self.witness_cols
+            .iter()
+            .filter(|(_, col)| col.stage <= self.stage as u32)
+            .map(|(poly_id, _)| poly_id)
+    }
+
+    fn polynomial_references(
+        &self,
+        expr: &impl AllChildren<AlgebraicExpression<T>>,
+    ) -> HashSet<PolyID> {
```

Comment:
Is it fine for this to be non-deterministic? Maybe `BTreeSet` is safer?

Reply:
It shouldn't matter, but I can change it!

Reply:
Hmm, it is used by the machine extractor, which uses hash sets everywhere. It would propagate quite far, so unless you feel strongly about it, I'd leave it as is. As long as its only used as a set, it should be fine.

Reply:
Not strongly, all good.